### PR TITLE
Feat/reward summarization example

### PR DESCRIPTION
**Pull request by TristanThrush**
This PR adds a summarization reward-model tuning example, following the "Learning to Summarize from Human Feedback" [paper](https://arxiv.org/abs/2009.01325).

### CODE REVIEW

File: examples/summarization/scripts/reward_summarization.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+    for info, summaries, choice in zip(examples["info"], examples["summaries"], examples["choice"]):
+        assert len(summaries) == 2
+        assert choice in (0,1)
+        original_text_field = "post" if info["post"] is not None else "article"
+        new_examples["text_j"].append(summaries[choice]["text"] + " " + tokenizer.bos_token + " " + info[original_text_field])
+        new_examples["text_k"].append(summaries[0 if choice==1 else 1]["text"] + " " + tokenizer.bos_token + " " + info[original_text_field])
+
+    return new_examples
+
+num_proc = 100 # Can adjust to be higher if you have more processors.
```
Comment (natolambert) [inline examples/summarization/scripts/reward_summarization.py:114]: Is this for different distributed training architectures?

**Review by natolambert — COMMENTED** [review]
Added some small comments (a somewhat quick read through, I can re-visit):
* Did you train a model, is it on the hub?
* You mentioned the results match the original paper, can you add that info to the PR?
* Can you pull main and run the linting on examples? `Make style`

File: examples/summarization/scripts/reward_summarization.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        attention_mask_k=None,
+        return_loss=True,
+    ):
+        transformer_outputs_j = self.transformer(
+            input_ids_j,
+            attention_mask=attention_mask_j,
+        )
+        hidden_states_j = transformer_outputs_j[0][:,-1,:]  # Get the last hidden state
+        rewards_j = self.v_head(hidden_states_j)
+        if input_ids_k is not None:
```
[the remark this answers is not in the document]
Comment (TristanThrush) [reply examples/summarization/scripts/reward_summarization.py:66]: Already in the comment at the top of the class. Can add above this line though

File: examples/summarization/scripts/reward_summarization.py (hunk 1 above)
Comment (TristanThrush) [reply examples/summarization/scripts/reward_summarization.py:114]: no it's just an argument that you can add in the datasets map function https://huggingface.co/docs/datasets/process

It should work even if your system doesn't have 100 cpus

File: examples/summarization/scripts/reward_summarization.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+    gpt_model_name: Optional[str] = field(default="gpt2", metadata={"help": "The gpt model that you want to train from the Hugging Face hub. E.g. gpt2, gpt2-xl, etc."})
+    bf16: Optional[bool] = field(default=False, metadata={"help": "This essentially cuts the training time in half if you want to sacrifice a little precision and have a supported GPU."})
+
+parser = HfArgumentParser(ScriptArguments)
+script_args = parser.parse_args_into_dataclasses()[0]
+
+# Load the human comparisons dataset for tuning the reward model.
+ds = load_dataset("openai/summarize_from_feedback", name="comparisons")
+
+# Define the reward model. Code here is loosely inspired by https://wandb.ai/carperai/summarize_RLHF/reports/Implementing-RLHF-Learning-to-Summarize-with-trlX--VmlldzozMzAwODM2.
```
[the remark this answers is not in the document]
Comment (TristanThrush) [reply examples/summarization/scripts/reward_summarization.py:34]: Yes I can write a docstring! What variables are you referring to.

File: examples/summarization/scripts/reward_summarization.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (natolambert) [reply examples/summarization/scripts/reward_summarization.py:66]: Yeah, a lesson from diffusers / open source: repeat yourself, makes people less confused.

File: examples/summarization/scripts/reward_summarization.py (hunk 1 above)
Comment (natolambert) [reply examples/summarization/scripts/reward_summarization.py:114]: Cool, maybe add a comment saying that (good plug to HF tools)

**Review by lvwerra — COMMENTED** [review]
Hi @TristanThrush, this is great - thanks for adding! I looked at the code and was thinking about ways to simplify it a bit further. Here's my proposal:

For the reward model we can use the vanilla `AutoModelForSequenceClassification`. This is also just a linear layer on top of the hidden states. Instead of having a custom model we could then just write a `compute_loss` for the `Trainer` (see `CustomTrainer` example [here](https://huggingface.co/docs/transformers/v4.26.0/en/main_classes/trainer#trainer)).

The advantage of this: 1) simplifies the code a bit and 2) we don't need a custom model class. For 2) this also allows you to use all the `.push_to_hub` and `from_pretrained` functionality to share the reward model without any changes.

If we are already having a custom `compute_loss` function we could just tokenize the dataset beforehand and pad to the max. Inside `compute_loss` we can remove the excessive padding from the batch (I think this could be a one-liner) before doing the forward pass. Then we wouldn't need a custom collator neither and the tokenization could be integrated into the `turn_into_text_classification_format`. 

Regarding the config: since we are not explaining exactly how to use it we could omit it here and refer to the DS integration of the Trainer?

What do you think?

File: examples/summarization/scripts/reward_summarization.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        model = AutoModelForCausalLM.from_pretrained(config)
+        self.config = model.config
+        # gpt-neo models have hidden_size instead of n_embd.
+        self.config.n_embd = (
+            self.config.hidden_size
+            if hasattr(self.config, "hidden_size")
+            else self.config.n_embd
+        )
+        self.transformer = model.transformer
+        self.v_head = nn.Linear(self.config.n_embd, 1)
```
[the remark this answers is not in the document]
Comment (TristanThrush) [reply examples/summarization/scripts/reward_summarization.py:51]: There are other reasonable reward models that we could imagine. Like, ones that return a softmax. Not sure if we should write this.

File: examples/summarization/scripts/reward_summarization.py [hunk 5]
```diff
@@ -40,13 +40,14 @@ def __init__(self, config):
         super().__init__()
         model = AutoModelForCausalLM.from_pretrained(config)
         self.config = model.config
-        # gpt-neo models have hidden_size instead of n_embd.
+        # Optional internal config change: gpt-neo models use hidden_size instead of n_embd.
```
Comment (TristanThrush) [inline examples/summarization/scripts/reward_summarization.py:43]: What is an "optional internal config change"?

**Comment by TristanThrush** [comment]
> Hi @TristanThrush, this is great - thanks for adding! I looked at the code and was thinking about ways to simplify it a bit further. Here's my proposal:
> 
> For the reward model we can use the vanilla `AutoModelForSequenceClassification`. This is also just a linear layer on top of the hidden states. Instead of having a custom model we could then just write a `compute_loss` for the `Trainer` (see `CustomTrainer` example [here](https://huggingface.co/docs/transformers/v4.26.0/en/main_classes/trainer#trainer)).
> 
> The advantage of this: 1) simplifies the code a bit and 2) we don't need a custom model class. For 2) this also allows you to use all the `.push_to_hub` and `from_pretrained` functionality to share the reward model without any changes.
> 
> If we are already having a custom `compute_loss` function we could just tokenize the dataset beforehand and pad to the max. Inside `compute_loss` we can remove the excessive padding from the batch (I think this could be a one-liner) before doing the forward pass. Then we wouldn't need a custom collator neither and the tokenization could be integrated into the `turn_into_text_classification_format`.
> 
> Regarding the config: since we are not explaining exactly how to use it we could omit it here and refer to the DS integration of the Trainer?
> 
> What do you think?

Thanks leandro, makes sense! I'm getting roped into the rlhf mturk stuff right now but I will try to address later in the week.

**Comment by TristanThrush** [comment]
> Added some small comments (a somewhat quick read through, I can re-visit):
> 
> * Did you train a model, is it on the hub?
> * You mentioned the results match the original paper, can you add that info to the PR?
> * Can you pull main and run the linting on examples? `Make style`
> * Should we add a small page in docs?

Thanks! These suggestions make sense to me. Will try to address later in the week!

File: examples/summarization/scripts/reward_summarization.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TristanThrush) [reply examples/summarization/scripts/reward_summarization.py]: I actually took this whole class out for my response to leandro's comment

File: examples/summarization/scripts/reward_summarization.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TristanThrush) [reply examples/summarization/scripts/reward_summarization.py]: I actually took this class out entirely

File: examples/summarization/scripts/reward_summarization.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (TristanThrush) [reply examples/summarization/scripts/reward_summarization.py]: took the class out entirely, per leandro's advice

File: examples/summarization/scripts/reward_summarization.py (hunk 1 above)
Comment (TristanThrush) [reply examples/summarization/scripts/reward_summarization.py]: done!

File: examples/summarization/scripts/reward_summarization.py (hunk 5 above)
Comment (TristanThrush) [reply examples/summarization/scripts/reward_summarization.py]: (actually this part of the code was removed anyway)

**Comment by TristanThrush** [comment]
Alright @lvwerra thanks again for your useful comments! In response, I:

* used AutoModelForSequenceClassification instead
* used a custom trainer with a custom `compute_loss` function
* I tried to not use a custom data collator, but I found that it was difficult for 2 reasons:
    1) I needed to specify `"return_loss": True` somewhere in the batched input data. If I don't do this, then `AutoModelForSequenceClassification` will error out, because my custom `compute_loss` function will not be used for the validation step. The custom `compute_loss` function needs to be used for training and validation instead of the normal model `forward` function. The normal `forward` function expects `input_ids`, not `input_ids_j` and `input_ids_k`. It is fine to use the normal `forward` function after the reward model is trained when you want to use it in the ppo stage, though.
    2) It seems a bit slower to pad to the max and then remove padding
    
I'm happy to try to make it work though if you still think that a custom data collator is not good!

**Comment by TristanThrush** [comment]
Thanks @natolambert for the comments. In response, I linted my code and added the doc pages. The one thing I havent addressed is pointing to trained model which matches the paper's results. I will do so soon, but am still training a model on the newly altered code.

File: examples/summarization/scripts/reward_summarization.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+# Need to do this for gpt2, because it doesn't have an official pad token.
+tokenizer.pad_token = tokenizer.eos_token
+model.config.pad_token_id = tokenizer.eos_token_id
+
+# Turn the dataset into pairs of post + summaries, where text_j is the preferred post + summary and text_k is the other.
+def turn_into_text_classification_format(examples):
+    new_examples = {"text_j": [], "text_k": []}
+    for info, summaries, choice in zip(examples["info"], examples["summaries"], examples["choice"]):
+        assert len(summaries) == 2
+        assert choice in (0, 1)
```
Comment (lvwerra) [inline examples/summarization/scripts/reward_summarization.py:89]: Can we do proper errors with useful messages? :)
```suggestion
        if len(summaries) != 2 or choice not in (0, 1):
            raise ValueError(f"There should be two summaries with a choice that's either 0 or 1. Received {len(summaries} summaries and choice={choice}.")
```

File: examples/summarization/scripts/reward_summarization.py [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+# Load the human comparisons dataset for tuning the reward model.
+ds = load_dataset("openai/summarize_from_feedback", name="comparisons")
+
+# Define the training args. Needs to be done before the model is loaded if you are using deepspeed.
+training_args = TrainingArguments(
+    output_dir=f"{script_args.model_name}_summarization_reward_model",
+    learning_rate=script_args.learning_rate,
+    per_device_train_batch_size=script_args.per_device_train_batch_size,
+    per_device_eval_batch_size=script_args.per_device_eval_batch_size,
+    num_train_epochs=5,
```
Comment (lvwerra) [inline examples/summarization/scripts/reward_summarization.py:63]: That seems like a parameter someone might also want to tune, right? so maybe move it to the arguments as well.

File: examples/summarization/scripts/reward_summarization.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    weight_decay=0.001,
```
Comment (lvwerra) [inline examples/summarization/scripts/reward_summarization.py:64]: same here probably

File: examples/summarization/scripts/reward_summarization.py [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+def compute_metrics(eval_pred):
+    predictions, _ = eval_pred
+    # Here, predictions is rewards_j and rewards_k.
+    # We want to see how much of the time rewards_j > rewards_k.
+    predictions = np.argmax(predictions, axis=0)
+    labels = np.zeros(predictions.shape)
+    return accuracy.compute(predictions=predictions, references=labels)
+
+
+class RewardTrainer(Trainer):
```
Comment (lvwerra) [inline examples/summarization/scripts/reward_summarization.py:174]: Maybe at some point we can a general `RewardTrainer` to TRL which could also be used to train on the ranking task!

File: examples/summarization/scripts/reward_summarization.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+        )
+        new_examples["text_k"].append(
+            summaries[0 if choice == 1 else 1]["text"] + " " + tokenizer.bos_token + " " + info[original_text_field]
+        )
+
+    return new_examples
+
+
+num_proc = (
+    100
```
Comment (lvwerra) [inline examples/summarization/scripts/reward_summarization.py:102]: it looks like a very quick operation so going for 100 processes might actually be more overhead than if we went maybe for 8?

**Review by lvwerra — COMMENTED** [review]
Awesome, thanks for the refactoring. Looks much cleaner now IMO!

Left just a few minor things to clean up then we can merge! 🚀

**Comment by natolambert** [comment]
Random comment on reward model, make sure we aren't assigning reward / gradients to the prompt / instruction tokens.

**Comment by lvwerra** [comment]
@natolambert you mean during RL training? The PPOTrainer takes care of that and only looks at rewards for the response tokens. For the RM training it doesn't really matter as we predict the reward after the last token, right?

File: examples/summarization/scripts/reward_summarization.py (hunk 7 above)
Comment (TristanThrush) [reply examples/summarization/scripts/reward_summarization.py:64]: Yeah really good point. 5 is the number of epochs that openai used, but it makes sense for me to make it an option and just make 5 the default. And then put that in the documentation.

File: examples/summarization/scripts/reward_summarization.py (hunk 10 above)
Comment (TristanThrush) [reply examples/summarization/scripts/reward_summarization.py:103]: yeah good point. The dataset size is fixed, after all.

**Comment by TristanThrush** [comment]
> Awesome, thanks for the refactoring. Looks much cleaner now IMO!
> 
> Left just a few minor things to clean up then we can merge! 🚀

Ok I believe I've addressed the comments. Will also add a model card to the linked model. Thanks, and let me know if there is anything else!
